internal/liveness/metrics.go

```diff
@@ -0,0 +1,252 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package liveness
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ceph/ceph-csi/internal/cephfs"
+	"github.com/ceph/ceph-csi/internal/rbd"
+	"github.com/ceph/ceph-csi/internal/util"
+
+	"github.com/kubernetes-csi/csi-lib-utils/rpc"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+var secrets = make(map[string]*util.Credentials)
+
+var (
+	rbdMapping = *prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "csi",
+		Name:      "ceph_rbd",
+		Help:      "ceph rbd image pvc mapping"},
+		[]string{"pvc_name", "pvc_namespace", "pv_name", "pool", "imageID"})
+	cephfsMapping = *prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "csi",
+		Name:      "ceph_fs",
+		Help:      "cephfs subvolume pvc mapping"},
+		[]string{"pvc_name", "pvc_namespace", "pv_name", "volume_name", "subvolume_group", "subvolume_name"})
+)
+
+const (
+	// csiConfigFile is the location of the CSI config file
+	csiConfigFile = "/etc/ceph-csi-config/config.json"
+)
+
+func getMapping(timeout time.Duration, client *k8s.Clientset, numGoroutines int, driverName string) {
+	util.DebugLogMsg("Start recording mapping")
+	pvcs, err := client.CoreV1().PersistentVolumeClaims("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf("Get pvclist failed: %v", err)
+		return
+	}
+	var relatedPvcs []v1.PersistentVolumeClaim
+	var length = 0
+	for index := range pvcs.Items {
+		var related bool = false
+		for key, value := range pvcs.Items[index].Annotations {
+			if strings.Contains(key, "storage-provisioner") && value == driverName {
+				related = true
+			}
+		}
+		if !related {
+			continue
+		}
+		relatedPvcs = append(relatedPvcs, pvcs.Items[index])
+		length++
+	}
+	var wg sync.WaitGroup
+	jobs := make(chan v1.PersistentVolumeClaim, numGoroutines)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*timeout)
+	defer cancel()
+	wg.Add(numGoroutines)
+	for i := 0; i < numGoroutines; i++ {
+		go getPvcMappings(ctx, client, jobs, &wg)
+	}
+	for index := range relatedPvcs {
+		jobs <- relatedPvcs[index]
+	}
+	close(jobs)
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		wg.Wait()
+	}()
+
+	select {
+	case <-done:
+		util.DebugLogMsg("Collect mapping metrics complete")
+	case <-time.After(timeout):
+		klog.Errorf("Collect mapping metrics timeout")
+	}
+}
+
+func collectPvcMapping(client *k8s.Clientset, pvc *v1.PersistentVolumeClaim, secrets map[string]*util.Credentials) {
```

Comment:
this function is pretty large. Add some empty lines between sections/tasks and comment the parts, please